### PR TITLE
Assembly: Correctly build reference from selection subname

### LINKED ISSUE #16207: Assembly: Origin moves causing visual jump
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

Assemblies break on small changes due to TNP.

In this assembly the cylinder in the middle is fixed to the hole of the first body:

Adding a chamfer to the sketch of the first body breaks the assembly:

Test file:
[assembly_tnp.FCStd.zip](https://github.com/user-attachments/files/16833873/assembly_tnp.FCStd.zip)

/cc @PaddleStroke

### Full version info

```shell
OS: macOS 14.6.1
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.38555 (Git)
Build type: Release
Branch: main
Hash: 73cb5bafe0c32c178d9b594a6ca03124483d305f
Python 3.11.9, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: C/Default (C)
Installed mods: 
  * OpenDark 2024.3.13
  * Manipulator 1.5.7
  * fasteners 0.5.28


----

Also happens in latest Ondsel build:
OS: macOS 14.6.1
Word size of Ondsel: 64-bit
Version: 2024.3.0dev.38472 (Git)
Build type: Release
Branch: flavor
Hash: 60c383a71ad9a470a811ccf6537156e03f180539
Python 3.11.9, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: C/Default (C)
Installed mods: 
  * Ondsel-Lens 2024.7.5.02
  * fasteners 0.5.24
  * Curves 0.6.41
```

### Subproject(s) affected?

Assembly

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by bgbsww**
Fix #16207.   I have switched from a python based solution to a lower level c++ one.

### CODE REVIEW

**Comment by bgbsww** [comment]
The code here is a workaround for the wrong subname/element coming back in the addSelection SelectionObserver callback.  That is ultimately coming from the PickedInfo in SoFCUnifiedSelection.cpp which needs extension / correction to handle TNP correctly.  Unfortunately, that is hundreds of lines of changes directly affecting selection, so needs careful consideration.

File: src/Mod/Assembly/JointObject.py [hunk 1]
```diff
@@ -1800,19 +1800,22 @@ def getMovingPart(self, ref):
...
+        # Alternatively we could take just the element and build back up using the sub_sub_element name to get a
+        # different legit path, but the first approach is preferable.
+        #   element_name, *path_detail = sub_name.split(".")
+        #   target_link = ".".join((element_name,sub_sub_element_name,old_name))
+        # Fixme: This code is a workaround that will break when the underlying code related to PickedInfo in
+        #  SoFCUnifiedSelection.cpp is corrected to process subname/element properly.
+        #  A review of the 13 places we define an addSelection reveals that the 'sub_name' (or 'element') parameter
+        #  would ne okay if those changes were in place.
+        element_name = sub_name.replace(sub_sub_element_name, "").replace(new_name, "")
+        target_link = element_name[:-1] + old_name
```
Comment (CalligaroV) [inline src/Mod/Assembly/JointObject.py:1817]: I'd say that we should improve this part of the code adding a check for the non-Body objects (for example, a nut made with Fasteners WB or other objects that don't have a structure like `Body.Feature.Element`)

A quick-and-dirty idea may be something like

```python
        element_name = (sub_name.replace(sub_sub_element_name, "") if "Body" in sub_name else sub_name).replace(new_name, "")
        target_link = (element_name[:-1] if "Body" in sub_name else element_name) + old_name
```

**Review by CalligaroV — CHANGES_REQUESTED** [review]
This modification surely looks a step forward, at least AFAICT after various local tests.

There's a modification I'd like to suggest, came out while testing the file provided by @felixscheibe in https://github.com/FreeCAD/FreeCAD/issues/16207#issuecomment-2354964550:
In that file there are some fasteners made with Fasteners WB and the modification in this PR strips the first part of those objects' names while trying to manually recreate the joint, not allowing to use any of that object's elements

One last thing, that may be discussed elsewhere, is that probably it could be a good thing to add some logic in `Joint.createProperties()` to automatically "clean" the names of the elements used (a possible new `Joint.migrationScript4()` method?) so that the user don't have do to that manually recreating the various joints.

File: src/Mod/Assembly/JointObject.py (hunk 1 above)
Comment (bgbsww) [reply src/Mod/Assembly/JointObject.py:1817]: I am probably not going to progress this, because none of this belongs at python level - that's a fundamental design error in the code, as noted in the comment.

The correct place to address this is in SoFCUnifiedSelection.cpp, in getFullSubElementName in lieu of a full rework.

I'm hoping there is a more generic solution, which is "is the elementmap of this element the same as the element map of the preceding element in the subelement path?"

Comment (bgbsww) [reply src/Mod/Assembly/JointObject.py]: Well, instead I commented out all the python level stuff and added the c++ pieces.  I would appreciate an additional review.

**Comment by bgbsww** [comment]
While I'm taking this out of draft, I would appreciate additional reviews here.

**Comment by yorikvanhavre** [comment]
waiting for reviews

**Comment by bgbsww** [comment]
Thanks for catching these.  I have a lot of ideas about how to narrow this, and the examples you provide will help me to work out what I can use.  Unfortunately I'm in a Gui context and don't want to create any dependencies into a Mod so I have to be careful about how to accomplish the filtering.  More to come.

**Comment by bgbsww** [comment]
I could not recreate your first issue with the new Body and new Sketch, but I certainly could replicate the Fillet.  This made it clear that I was overapplying the filter; it should only attempt to shorten names that have at least three components; the  container ( Body, Part, etc ), the Object ( Pad, etc ), and the element.  Thus the change to ensure at least two dots in the name for it to be eligible.  Let me know if this works for your cases, please.

**Review by CalligaroV — COMMENTED** [review]
I can say pretty confidently that we're moving forward with this.

Trying with
```ruby
OS: Linux Mint 21.3 (X-Cinnamon/cinnamon)
Word size of FreeCAD: 64-bit
Version: 1.1.0dev.38797 +1 (Git)
Build type: Debug
Branch: bgbsww-AssemblySubNames
Hash: f0053e5388c96b0d2bf80b456bbae3ac0901ca7a
Python 3.10.12, Qt 5.15.3, Coin 4.0.0, Vtk 7.1.1, OCC 7.8.1
Locale: Italian/Italy (it_IT)
Stylesheet/Theme/QtStyle: OpenLight.qss/OpenLight/Qt default
Installed mods: 
  * Assembly4 0.50.15
  * freecad.gears 1.2.0
  * Beltrami 1.1.0
  * sheetmetal 0.4.24
  * Curves 0.6.44
  * fasteners 0.5.26
  * kicadStepUpMod 11.2.7
  * OpenTheme 2024.8.10
  * Render 2024.8.4
  * Ondsel-Lens 2024.7.5.02
```

I see that my previous issue with the Fillet is solved, as well as some of the examples of the original issue that led to this PR.

I'd really like to submit an Approval feedback for this PR but there are still a couple of things I found while testing this PR that should be fixed, otherwise there will be regressions:

I'm still experiencing the issue of the selection of a Base plane to sketch on in an empty Body.
I guess we're doing the same steps, but to avoid misunderstanding here are the ones I take to trigger this issue:
 - Start FreeCAD in GUI mode
 - In the Start Page press `Parametric Part` (this should create a new document with an empty Body and automatically activate the PartDesign WB)
 - From the Task Panel press `Create sketch`
 - From the 3D View select a Base plane (doesn't matter which one)
    - **Expected**: you can start sketching (auto switch to Sketcher WB)
    - **Actual** (at least for me): the entire Origin gets selected

100% reproducible for me

Another difference I noticed between this PR and

```ruby
OS: Linux Mint 21.3 (X-Cinnamon/cinnamon)
Word size of FreeCAD: 64-bit
Version: 1.1.0dev.38841 (Git)
Build type: Debug
Branch: main
Hash: 9df2ec7b86aeecc066c0fe703b62fa5780eb3374
Python 3.10.12, Qt 5.15.3, Coin 4.0.0, Vtk 7.1.1, OCC 7.8.1
Locale: Italian/Italy (it_IT)
Stylesheet/Theme/QtStyle: OpenLight.qss/OpenLight/Qt default
Installed mods: 
  * Assembly4 0.50.15
  * freecad.gears 1.2.0
  * Beltrami 1.1.0
  * sheetmetal 0.4.24
  * Curves 0.6.44
  * fasteners 0.5.26
  * kicadStepUpMod 11.2.7
  * OpenTheme 2024.8.10
  * Render 2024.8.4
  * Ondsel-Lens 2024.7.5.02
``` 
is that:
 - **In main**: if there's a Body inside an `Std::Part`, selecting it's shape from the 3D View and then performing a Delete operation (for example with the Delete button on the keyboard), then the `Feature` related to the selected shape is deleted - 
 - **In this PR**: with the same scenario of above, performing a Delete operation (for example with the Delete button on the keyboard) deletes the entire Body (and all that's left is a Part with it's origin)

Also this one is 100% reproducible for me

As for my previous comment, the first thing I'd like to ask is if those issues are experienced only by me (and AFAICS it looks so at least for the selection of a Base plane)

If so then I haven't anything against merging this PR.

**Comment by bgbsww** [comment]
Yeah, the first one with the base plane of a sketch I absolutely can't make happen - it just works normally.

On the second one, if I watch the tree and the selection view, and double click to rotate through selection states, then
everything goes as I would expect:  If the inner Feature ( Pad of a Sketch ) is selected, the Pad is deleted and the Sketch stays.  If the Body is selected, the Body, Pad, and Sketch are removed.  If the Part is selected, delete brings up a dialog asking whether to delete the contents as well; both options do as expected.

When I try the same thing in the daily,  I get the same behaviors.  I am suspicious that something is funny in your build that is affecting both things, or maybe you don't actually have the latest of this PR - maybe that's a side effect of the force-pushing?

OS: Ubuntu 24.04.1 LTS (ubuntu:GNOME/ubuntu)
Word size of FreeCAD: 64-bit
Version: 1.1.0dev.38834 (Git)
Build type: Release
Branch: main
Hash: 490d6c5abc41bb61fa5bf47398df142a8ac2519e
Python 3.12.3, Qt 5.15.13, Coin 4.0.2, Vtk 9.1.0, OCC 7.8.1
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: FreeCAD Dark.qss/FreeCAD Dark/Qt default
Installed mods: 
  * OpenTheme 2024.9.1
  * sheetmetal 0.4.26
  * Curves 0.6.47
  * DynamicData 2.63.0
  * BIM021 2021.12.0
  * parts_library

**Comment by bgbsww** [comment]
There isn't a non-hacky way to do this, which means I need to leave the selection code alone at the c++ layer.

Then the question becomes one of what the correct thing is at the Assembly python layer.  Trying to deal with this as TNP names is clearly wrong, but maybe I let myself be confused by the comments in there.

What is what we're trying to say is that Assembly should be moving references up one layer from Features to the Body that contains them to avoid a problem?  I'll take a shot at that.

**Comment by bgbsww** [comment]
Okay.  Return to python, get rid of anything TNP, clean the names by simply removing the second to last piece when building a joint, and take @CalligaroV suggestion to add a migration so that existing files work too.

Fingers crossed I think this is finally it.

**Comment by bgbsww** [comment]
Looks like conda builds are broken again :(

**Review by CalligaroV — APPROVED** [review]
This is very good!!!

I did quite some tests in the past couple of days and AFAICS the modifications done so far surely help making Assembly WB more robuts against TNP.

Tested (needed to rebase to leverage also on other merged PRs, for example the one related to `PartDesign::Hole`):
```ruby
OS: Linux Mint 21.3 (X-Cinnamon/cinnamon)
Word size of FreeCAD: 64-bit
Version: 1.1.0dev.38843 +3 (Git)
Build type: Debug
Branch: bgbsww-AssemblySubNames
Hash: 2c04b7a838d1cc174ec8659f4495ef23a2e74618
Python 3.10.12, Qt 5.15.3, Coin 4.0.0, Vtk 7.1.1, OCC 7.8.1
Locale: Italian/Italy (it_IT)
Stylesheet/Theme/QtStyle: OpenLight.qss/OpenLight/Qt default
Installed mods: 
  * Assembly4 0.50.15
  * freecad.gears 1.2.0
  * Beltrami 1.1.0
  * sheetmetal 0.4.24
  * Curves 0.6.44
  * fasteners 0.5.26
  * kicadStepUpMod 11.2.7
  * OpenTheme 2024.8.10
  * Render 2024.8.4
  * Ondsel-Lens 2024.7.5.02
``` 
against:
[assembly_tnp.FCStd.zip](https://github.com/user-attachments/files/16833873/assembly_tnp.FCStd.zip)
[assembly_INT.zip](https://github.com/user-attachments/files/17019874/assembly_INT.zip)
[AssemplyTNPIssue.zip](https://github.com/user-attachments/files/17025533/AssemplyTNPIssue.zip)
[Assembly-WB_Filet Problem.zip](https://github.com/user-attachments/files/17006339/Assembly-WB_Filet.Problem.zip)

This is what I found:
 - [assembly_tnp.FCStd.zip](https://github.com/user-attachments/files/16833873/assembly_tnp.FCStd.zip) works as expected by the user that opened the issue. It also works if modifications are made with new features. Issues may happen if multiple modifications are made both in the original sketch and in newer features, but that is more a PD/TNP topic then an Assembly one.
 - [assembly_INT.zip](https://github.com/user-attachments/files/17019874/assembly_INT.zip) is probably the most troublesome: adding the chamfer to the edge specified by the user always moves the other part involved in the joint, regardless for modifications done to previous features or newer ones. I suspect however that this is caused by something missing in `PartDesign::Chamfer`
 - [AssemplyTNPIssue.zip](https://github.com/user-attachments/files/17025533/AssemplyTNPIssue.zip) Works as expected by the user. Issues jump out if modifications are made to the already existing `PartDesign::Pad` and/or `PartDesign::Hole`. Looks like something misses notifying to `App.ActiveDocument.Revolute.Reference[1/2]` that some names in the references have been updated (no `[...] auto change element reference [...]` message)
 - [Assembly-WB_Filet Problem.zip](https://github.com/user-attachments/files/17006339/Assembly-WB_Filet.Problem.zip) No visual jumps occurs if the modification is made as specified by the user. Some joints however may not work properly: in this case a Parallel joint loses two references  (a Face and a Vertex) because the Chamfer deletes them (I think that #16803 may help with that). Regardless for this, other modifications are handled correctly similarly to what happens with [assembly_tnp.FCStd.zip](https://github.com/user-attachments/files/16833873/assembly_tnp.FCStd.zip)

As always, this is what I can say as result of my tests and I would be glad if someone else could perform some, there may be something in my build that gives false positives/negatives.

There's only one small modification that I'd like to ask you @bgbsww:
in https://github.com/FreeCAD/FreeCAD/blob/d535c5a6e7833837bd11d8a6584a580a9eaad128/src/Mod/Assembly/JointObject.py#L1841
I think it should be more correct to replace `sub_name` with `target_link`, otherwise users won't be able anymore to remove the references of a joint by clicking on the related element in the 3D View (as happens right now in main).
TBH, adding this modification doesn't fully restores main's behavoir (the first click doesn't have visible effect) but allows again to modify the joint's references through the 3D View

---
**Summarizing: (AFAICT)**
Even if this PR doesn't fully fixes all the cases described in the linked issue, IMHO it's good to merge and actually makes Assembly more robust/aware of TNP.

Nice to have, by the UX point of view, the modification I suggested.

Cases not handled 100% correctly should be (are already) discussed in other issues/PRs.
In any case, to me this looks actually fixing #16207

**Comment by maxwxyz** [comment]
@CalligaroV I'll mark it ready for review then.

**Comment by chennes** [comment]
Backported to 1.0

**Comment by PaddleStroke** [comment]
It appears that this PR is actually breaking some things in assembly. As this processing of the subname works if you select Part::Features of a body. But it is breaking if you select datums.

I'd like to understand clearly what this PR is doing so that I don't break it back the other way around. @CalligaroV could you please confirm if I got this right?

The goal of this PR is to remove the name of the body's feature. ie instead of  `Part.Body.Pad.Sketch` we have ` Part.Body.Sketch`.
Also if you select a sketch in the tree, it appears that the name returned by selection is `Part.Body.Pad.Sketch.` which is not correct. As if you select in the view you get `Part.Body.Sketch.`

Basically this is removing the body's feature name right?

**Comment by PaddleStroke** [comment]
Example where it breaks : 
Select one edge of a sketch like : 

        print(sub) ->Part.Body.Sketch001.;g2;SKT.Edge2
        print(base_name) ->Part
        print(*path_detail) -> Body Sketch001 ;g2;SKT
        print(old_name) -> Edge2
        print(target_link) -> Part.Body.Edge2
        print(ref) -> [<Assembly object>, ['Part.Body.Edge2']]

Result it registers Body.Edge2

Example2 : 
        print(sub) ->Part.Body.Origin.Y_Axis.
        print(base_name) ->Part
        print(*path_detail) -> Body Origin Y_Axis
        print(old_name) -> 
        print(target_link) -> Part.Body.
        print(ref) -> [<Assembly object>, ['Part.Body.']]

**Comment by PaddleStroke** [comment]
@CalligaroV was it intended that : `Part.Body.Pad.Edge2` becomes `Part.Body.Edge2` as well?

**Comment by PaddleStroke** [comment]
Another problem is that `MigrationScript4 `is running everytime you open a new document. And so everytime it is trimming the subname. This must be removed.

**Comment by maxwxyz** [comment]
@PaddleStroke is there an issue for this or a recent PR so this won't get lost?

**Comment by CalligaroV** [comment]
@PaddleStroke, yes, the original purpose of this PR was to trim the name (not the TNP-related cryptic element name) of a geometrical object because in some cases having the entire name, with also the PD operation, makes Assembly more weak against TNP changes.

I studied this PR few weeks ago and I'll need some time (luckly not much as this isn't too old) to study it again and understand better what it does and if there may be other ways to achieve the same result while not breaking anything.

I see that you already opened a new PR, therefor I think I'll add further comments there.
